### DISCUSSION TITLE
Export IfcGridAxis сoordinates with IfcConvert

### DISCUSSION THREAD

**Discussion by rosterplayer** [IfcOpenShell/IfcOpenShell#5401] [Q&A] [answered]
Good day!
I utilize the IFCConvert tool with the parameters "--model --plan" to facilitate the conversion of IFC files to .glb. However, I also require further information regarding the grid axes, in order to display the axis labels on my application. I have made minor modifications to ifcConvert.cpp and exported the data to JSON as follows:
`...
		auto castedAxis = dynamic_cast<typename Schema::IfcGridAxis*>(rel);
		auto axisCurve = castedAxis->AxisCurve();

		boost::optional<std::string> axisName = castedAxis->AxisTag();

		Handle(Geom_Curve) curve;

		IfcGeom::Kernel kernel(&f);

		gp_Pnt startPoint;
		gp_Pnt endPoint;

		TopoDS_Wire wire;

		if (kernel.convert_curve(axisCurve, curve))
		{
			auto start = curve->FirstParameter();
			auto end = curve->LastParameter();
			startPoint = curve->Value(start);
			endPoint = curve->Value(end);
		}
		else if (kernel.convert_wire(axisCurve, wire))
		{
#if OCC_VERSION_HEX < 0x70600
			BRepAdaptor_CompCurve cc(wire, true);
			Handle(Adaptor3d_HCurve) hcc = Handle(Adaptor3d_HCurve)(new BRepAdaptor_HCompCurve(cc));
#else
			auto hcc = new BRepAdaptor_CompCurve(w, true);
#endif

			auto start = hcc->FirstParameter();
			auto end = hcc->LastParameter();
			startPoint = hcc->Value(start);
			endPoint = hcc->Value(end);
		}
...
`
My issue that coordinates of startPoint and endPoint not the same(for some ifc models) that in exported in .glb geometry. In my app, the grids and labels are in different places.

Could you advise me what's wrong?

**Comment by aothms** [comment]
Hard to tell. Either IfcGrid.ObjectPlacement (the placement of the grid) the equivalent of a 4x4 matrix. Or the use of FirstParameter/LastParameter. Do you expect these axis positions in local coordinates within the grid placement or world coordinates?

Or, I would use TopExp::Vertices on the TopoDS_Wire instead and then use BRep_Tool::Pnt() to get their XYZ.

**Reply by rosterplayer** [reply]
As far as I understand world coordinates are used when exporting to .glb? So i expect woorld coordinates for axis too.

**Reply by aothms** [reply, marked as answer]
Then you need to convert the IfcGrid.ObjectPlacement  into a gp_Trsf (use the kernel for that as well) and gp_Pnt::Transformed(gp_Trsf&) to do the transformation.

**Reply by rosterplayer** [reply]
Thank you, that was very helpful!
